### PR TITLE
[shopsys] remove deprecations from 7.x.x

### PR DESCRIPTION
**Pull request by boris-brtan**
| Q             | A
| ------------- | ---
|Description, reason for the PR| since we are almost near the release of next major version, we need to simplify code that was implemented to apply backward compatibility promise. One deprecated thing we kept, it is CartController:addProductFormAction since it is still used in productDetail but we have a plan to reimplement it with the new CartController:productActionAction.
|New feature| No 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| Yes 
|Fixes issues|  
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Component/Elasticsearch/ElasticsearchStructureManager.php [hunk 1]
```diff
@@ -26,50 +30,17 @@ class ElasticsearchStructureManager
...
-
-    /**
+     * @param string $buildVersion
      * @param string $definitionDirectory
      * @param string $indexPrefix
      * @param \Elasticsearch\Client $client
-     * @param string|null $buildVersion
      */
-    public function __construct(string $definitionDirectory, string $indexPrefix, Client $client, ?string $buildVersion = null)
+    public function __construct(string $buildVersion, string $definitionDirectory, string $indexPrefix, Client $client)
```
Comment (PetrHeinz) [inline packages/framework/src/Component/Elasticsearch/ElasticsearchStructureManager.php]: Do we really want to change the order of parameters?

**Comment by boris-brtan** [comment]
rebased on master

File: packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php [hunk 2]
```diff
@@ -15,8 +16,13 @@
 use Shopsys\FrameworkBundle\Model\Product\ProductVisibility;
 use Shopsys\FrameworkBundle\Model\Product\ProductVisibilityRepository;
 
-class ProductSearchExportWithFilterRepository extends ProductSearchExportRepository
+class ProductSearchExportWithFilterRepository
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php]: in the method `createQueryBuilder` the parameter `$locale` is never used. I have a feeling it could be removed because it is in the method only to fulfill the parent signature. But it would also mean an upgrade instruction...

So the decision is up to you

File: project-base/tests/ShopBundle/Functional/Model/Product/Search/ProductSearchExportRepositoryTest.php [hunk 3]
```diff
@@ -1,80 +0,0 @@
-<?php
-
-namespace Tests\ShopBundle\Functional\Model\Product\Search;
-
-use Shopsys\FrameworkBundle\Component\Domain\Domain;
-use Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportRepository;
-use Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportWithFilterRepository;
-use Tests\ShopBundle\Test\TransactionFunctionalTestCase;
-
-class ProductSearchExportRepositoryTest extends TransactionFunctionalTestCase
```
Comment (simara-svatopluk) [inline project-base/tests/ShopBundle/Functional/Model/Product/Search/ProductSearchExportRepositoryTest.php]: If this test is useful, please don't remove it and rewrite it for the `ProductSearchExportWithFilterRepository`

File: packages/framework/src/Resources/config/services_test.yml [hunk 4]
```diff
@@ -234,9 +212,14 @@ services:
         arguments:
             $indexPrefix: 'test-%env(ELASTIC_SEARCH_INDEX_PREFIX)%'
 
+    Elasticsearch\Client:
```
Comment (simara-svatopluk) [inline packages/framework/src/Resources/config/services_test.yml]: Why are these changes here? Maybe I'm just surprised because the change is in the commit 86abff5780

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlDataFactory.php [hunk 5]
```diff
@@ -4,29 +4,6 @@
 
 class FriendlyUrlDataFactory implements FriendlyUrlDataFactoryInterface
 {
-    /**
-     * @deprecated since Shopsys Framework 7.3, use createFromIdAndName instead
-     *
-     * @param array $data
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    public function createFromData($data)
```
Comment (simara-svatopluk) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlDataFactory.php]: please add upgrade instruction for this class

File: packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php [hunk 6]
```diff
@@ -68,30 +67,4 @@ public function getRouteName(): string
...
         return static::ROUTE_NAME;
     }
-
-    /**
-     * @internal This method will be inlined when its implementation will be able to be simplified
-     *
-     * @param array $data
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    protected function createFriendlyUrlData(array $data): FriendlyUrlData
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Article/ArticleDetailFriendlyUrlDataProvider.php]: please add upgrade instruction for this class

File: packages/framework/src/Model/Product/Brand/BrandDetailFriendlyUrlDataProvider.php [hunk 7]
```diff
@@ -69,30 +68,4 @@ public function getRouteName(): string
...
         return static::ROUTE_NAME;
     }
-
-    /**
-     * @internal This method will be inlined when its implementation will be able to be simplified
-     *
-     * @param array $data
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    protected function createFriendlyUrlData(array $data): FriendlyUrlData
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Product/Brand/BrandDetailFriendlyUrlDataProvider.php]: please add upgrade instruction for this class

File: packages/framework/src/Model/Product/ProductDetailFriendlyUrlDataProvider.php [hunk 8]
```diff
@@ -71,30 +70,4 @@ public function getRouteName(): string
...
         return static::ROUTE_NAME;
     }
-
-    /**
-     * @internal This method will be inlined when its implementation will be able to be simplified
-     *
-     * @param array $data
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    protected function createFriendlyUrlData(array $data): FriendlyUrlData
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Product/ProductDetailFriendlyUrlDataProvider.php]: please add upgrade instruction for this class

File: packages/framework/src/Model/Product/ProductListFriendlyUrlDataProvider.php [hunk 9]
```diff
@@ -72,30 +71,4 @@ public function getRouteName(): string
...
         return static::ROUTE_NAME;
     }
-
-    /**
-     * @internal This method will be inlined when its implementation will be able to be simplified
-     *
-     * @param array $data
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    protected function createFriendlyUrlData(array $data): FriendlyUrlData
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Product/ProductListFriendlyUrlDataProvider.php]: please add upgrade instruction for this class

File: packages/framework/src/Model/Product/Search/Export/ProductSearchExporter.php [hunk 10]
```diff
@@ -47,87 +46,32 @@ class ProductSearchExporter
...
-            @trigger_error(sprintf('The %s() method is deprecated and will be removed in the next major. Use the constructor injection instead.', __METHOD__), E_USER_DEPRECATED);
-
-            $this->entityManager = $entityManager;
-        }
-    }
-
     /**
      * @param int $domainId
      * @param string $locale
-     * @deprecated Use `exportWithOutput` instead
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Model/Product/Search/Export/ProductSearchExporter.php]: hi @simara-svatopluk, i would remove this commit and let it for #1195 because it will be merged into master from 7.3
what do you think?

File: packages/framework/src/Resources/config/services_test.yml (hunk 4 above)
Comment (boris-brtan) [reply packages/framework/src/Resources/config/services_test.yml]: my fault during fixes for the tests i will remove it

File: packages/framework/src/Model/Product/Search/Export/ProductSearchExporter.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Model/Product/Search/Export/ProductSearchExporter.php]: Well if this method is going to be used in the future, remove the deprecation. I think this is true a correct current state. (meaning keep it as it is now :-) )

I just wanted to point it out, that's all. And I still think that the comment explaining the duplication could be helpful.

**Comment by simara-svatopluk** [comment]
We have a solution for deprecated export methods used by `ProductSearchExportCronModule`

```php
    public function run()
    {
        $symfonyStyle = new SymfonyStyle(new ArrayInput([]), new NullOutput());
        $this->productSearchExportFacade->exportAllWithOutput($symfonyStyle);
    }
```

So you can remove `ProductSearchExportFacade::exportAll()` and `ProductSearchExporter::export()`

**Comment by boris-brtan** [comment]
@simara-svatopluk i am not sure about AbstractMigration since there is some test for it and it was not done because of bc promise. What do you think?

**Comment by simara-svatopluk** [comment]
* I was thinking that "deprecation for unexpected type of price in session is removed" deserves upgrade instructions, but after short time I realized it was introduced in 7.0 and now it's time to calmly remove the code without instructions :heavy_check_mark: 
* AbstractMigration shouldn't be adressed, now I understand the reason `@deprecated` have to stay :heavy_check_mark: 
* you didn't remove deprecated methods `ProductSearchExportFacade::exportAll()` and `ProductSearchExporter::export()` as I suggested in previous comment. Any reason? :heavy_multiplication_x:

File: packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php [hunk 11]
```diff
@@ -123,6 +127,17 @@ public function getProductsDataForIds(int $domainId, string $locale, array $prod
         return $result;
     }
 
+    /**
+     * @param int $domainId
+     * @return int
+     */
+    public function getProductTotalCountForDomainAndLocale(int $domainId): int
```
Comment (simara-svatopluk) [inline packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php]: This method deserves to be renamed to `getProductTotalCountForDomain`, please don't forget upgrade notes

**Comment by boris-brtan** [comment]
OK so we will presume that in #1195 it will stay deprecated

File: packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php (hunk 11 above)
Comment (boris-brtan) [reply packages/framework/src/Model/Product/Search/Export/ProductSearchExportWithFilterRepository.php]: what about this solution ? https://github.com/shopsys/shopsys/pull/1193/commits/c7924abe35b49590167e38467fedec56eac060a2

File: docs/upgrade/UPGRADE-unreleased.md [hunk 12]
```diff
@@ -119,6 +119,50 @@ There you can find links to upgrade notes for other versions too.
...
+        - remove registration of `Shopsys\FrameworkBundle\Command\RedisCleanCacheOldCommand: ~` from `commands.yml`
+    - check whether you extended classes that constructor signature was changed and fix them
+        - `Shopsys\FrameworkBundle\Model\Mail\MailTemplateFacade`
+        - `Shopsys\FrameworkBundle\Component\Router\DomainRouterFactory`
+        - `Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExporter`
+        - `Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportStructureFacade`
+        - `Shopsys\FrameworkBundle\Model\Product\ProductFacade`
+        - `Shopsys\FrameworkBundle\Model\Product\ProductVariantFacade`
+        - `Shopsys\FrameworkBundle\Component\Elasticsearch\ElasticsearchStructureManager`
+    - check wether you have overwritten methods `createQueryBuilder` and renamed method `getProductTotalCountForDomainAndLocale` that is now named `getProductTotalCountForDomain` of `Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportWithFilterRepository` class and fix their signature
```
Comment (simara-svatopluk) [inline docs/upgrade/UPGRADE-unreleased.md]:
```suggestion
    - check wether you have overwritten methods of `Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportWithFilterRepository` class and fix their signatures
        - `createQueryBuilder` has less parameters
        - `getProductTotalCountForDomainAndLocale` was renamed to `getProductTotalCountForDomain` and has less parameters
```
